crypto/rand/drbg_lib.c

```diff
@@ -304,6 +304,41 @@ void rand_drbg_cleanup_nonce(RAND_DRBG *drbg,
     OPENSSL_clear_free(out, outlen);
 }
 
+
+
+/*
+ * Set the RAND_DRBG callback data for obtaining entropy and nonce.
+ *
+ * This is a specialized feature which is used for example to implement
+ * deterministic signatures according to RFC6979, which requires passing
+ * a specified entropy and nonce for instantiating the DRBG.
+ *
+ * The pointer is stored for later retrieval, but the ownership of the
+ * memory to which it points remains with the caller. (I.e., if the memory
+ * was allocated on the heap, then it is the caller's responsibility to
+ * free the memory after use.)
+ *
+ * Setting the callback data is allowed only if the drbg has not been
+ * initialized yet. Otherwise, the operation will fail.
+ *
+ * Returns 1 on success, 0 on failure.
+ */
+int rand_drbg_set_callback_data(RAND_DRBG *drbg, void *data)
+{
+    if (drbg->state != DRBG_UNINITIALISED
+        || drbg->parent != NULL)
+        return 0;
+
+    drbg->callback_data = data;
+    return 1;
+}
+
+/* Retrieve the pointer to the callback data */
+void *rand_drbg_get_callback_data(RAND_DRBG *drbg)
+{
+    return drbg->callback_data;
+}
```

Comment:
I am pondering whether I should make those two functions public: They are a useful supplement and enhancement of the callback mechanism. It could not only be used for deterministic signatures, but also for the known answer tests, which currently utilize the ex_data mechanism. The new API would simplify the implementation, because it does not require a RUN_ONCE initialization for the ex_data index.

```C
    int RAND_DRBG_set_callbacks(RAND_DRBG *drbg,
                                RAND_DRBG_get_entropy_fn get_entropy,
                                RAND_DRBG_cleanup_entropy_fn cleanup_entropy,
                                RAND_DRBG_get_nonce_fn get_nonce,
                                RAND_DRBG_cleanup_nonce_fn cleanup_nonce);

    int RAND_DRBG_set_callback_data(RAND_DRBG *drbg, void *data);
    void *RAND_DRBG_get_callback_data(RAND_DRBG *drbg);
```
@paulidale @slontis what do you think about my suggestion?

Reply:
I was looking at a different approach to this problem as part of #10144: the introduction of a test DRBG that output fixed data.  A DRBG could use this as its parent.  The context was for the FIPS known answer tests.

Unfortunately, #10144 seems like it is being too ambitious on other fronts.  It tried to split the DRBGs into legacy APIs, an EVP_RAND layer and implementations in providers.  It also tried to allow either an entropy source or a DRBG to be the parent of a DRBG.

As for making these public, I don't have a strong feeling either way.  We will need fetchable RANDs for FIPS, so not making that more difficult is a consideration but I don't see any (more) hurt happening because of this.

I'm not sure if it is relevant, but `OSSL_CALLBACK` includes an extra argument which could be used for this purpose.  They are a bit lower level though.

Reply:
The ex_data has been ripped out of FIPS_MODE so a mechanism using OSSL_CALLBACK would be a good idea.